src/python/zquantum/core/measurement.py

```diff
@@ -453,6 +453,50 @@ def get_expectation_value_from_frequencies(
     return expectation
 
 
+def _check_sample_elimination(
+    samples: collections.Counter,
+    bitstring_samples: List[Tuple[int, ...]],
+    leftover_distribution: BitstringDistribution,
+) -> collections.Counter:
+    """This is a recursive function, that checks that all elements in samples
+    are present in bitstring_samples. If they are not, we eliminate the
+    elements not in bitstring samples, set their probability to 0 in leftover_distribution
+    and resample the appropriate number of times. Then, we re-check the new samples.
+    Args:
+        samples: The bitstrings to eliminate and how many times to eliminate them
+        bitstring_samples: the bitstring distribution from where the bitstrings
+                           should be removed
+        leftover_distribution: the distribution from which samples to eliminate are
+                               sampled
+    Returns:
+        correct_samples: A sample object that only contains bitstrings that can actually
+                     be removed from bitstring_samples
+    """
+    bitstring_counts = Counter(bitstring_samples)
+    new_samples = None
+    correct_samples = samples
+    for sample in samples:
+        bitstring = tuple([int(measurement_value) for measurement_value in sample])
+        if samples[sample] > bitstring_counts[bitstring]:
+            nresamples = samples[sample] - bitstring_counts[bitstring]
+            samples[sample] = bitstring_counts[bitstring]
+            distribution_dict = leftover_distribution.distribution_dict
+            distribution_dict[sample] = 0
+            leftover_distribution = BitstringDistribution(distribution_dict, True)
+            new_samples = sample_from_probability_distribution(
```

Comment:
Could you help me understand why `new_samples` having any elements is the condition for recursion? 

Also, this might be easily changed to a `while` loop (in principle any tail recursion can).

Reply:
Fixed in 2a6f36d8f784948f037f5152374e935f1ec99216